### ISSUE TITLE
Cannot start syslog-ng with tls destination - Error setting up TLS session context; tls_error='(null):(null):(null)'

### ISSUE THREAD

**Issue by hydrapolic** [balabit/syslog-ng#1916]
# syslog-ng

## Version of syslog-ng

```
syslog-ng 3 (3.14.1)
Config version: 3.14
Installer-Version: 3.14.1
Revision: 
Compile-Date: Mar  7 2018 16:10:13
Module-Directory: /usr/lib64/syslog-ng
Module-Path: /usr/lib64/syslog-ng
Available-Modules: linux-kmsg-format,affile,stardate,tfgetent,tags-parser,syslogformat,csvparser,afsocket,cryptofuncs,graphite,xml,afstomp,disk-buffer,afprog,system-source,kvformat,map-value-pairs,basicfuncs,dbparser,afuser,appmodel,confgen,snmptrapd-parser,cef,date,add-contextual-data,pseudofile
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: off
Enable-TCP-Wrapper: on
Enable-Linux-Caps: off
Enable-Systemd: off
```

## Platform

Gentoo Linux

# Issue

## Failure

```
[2018-03-07T16:18:30.303624] Error setting up TLS session context; tls_error='(null):(null):(null)'
[2018-03-07T16:18:30.303658] Error initializing message pipeline; plugin name='network', location='/etc/syslog-ng/syslog-ng.conf:3:28'
```

## Steps to reproduce

1. start syslog-ng

## Configuration

```
@version: 3.14
source src { system(); internal(); };
destination syslog { network("syslog.example.com" port(1234) transport("tls") tls(ca-dir("/etc/ssl/certs") trusted-keys("SHA1:AA:BB:CC...") peer-verify("required-trusted")) ); };
log { source(src); destination(syslog); };
```

**Comment by hydrapolic** [comment]
The same configuration works with version 3.13.2.

**Comment by MrAnno** [comment]
Hi,

Thanks for reporting this issue.

Could you share a few more lines of your syslog-ng output preceding the line `Error setting up TLS session context...`?

**Comment by gaborznagy** [comment]
@hydrapolic!

I have reproduced the issue with the very same configuration, WITHOUT creating any certs or keys.
There is no other error messages on the command line before the reported error message, so we need to check our error handling!
Thanks for the finding!

**Comment by hydrapolic** [comment]
Full log in debug mode:

````
[2018-03-07T16:59:24.741350] Registering candidate plugin; module='linux-kmsg-format', context='format', name='linux-kmsg'
[2018-03-07T16:59:24.741380] Registering candidate plugin; module='affile', context='source', name='file'
[2018-03-07T16:59:24.741397] Registering candidate plugin; module='affile', context='source', name='pipe'
[2018-03-07T16:59:24.741433] Registering candidate plugin; module='affile', context='source', name='wildcard_file'
[2018-03-07T16:59:24.741458] Registering candidate plugin; module='affile', context='source', name='stdin'
[2018-03-07T16:59:24.741471] Registering candidate plugin; module='affile', context='destination', name='file'
[2018-03-07T16:59:24.741475] Registering candidate plugin; module='affile', context='destination', name='pipe'
[2018-03-07T16:59:24.741479] Registering candidate plugin; module='stardate', context='template-func', name='stardate'
[2018-03-07T16:59:24.741483] Registering candidate plugin; module='tfgetent', context='template-func', name='getent'
[2018-03-07T16:59:24.741487] Registering candidate plugin; module='tags-parser', context='parser', name='tags-parser'
[2018-03-07T16:59:24.741500] Registering candidate plugin; module='syslogformat', context='format', name='syslog'
[2018-03-07T16:59:24.741515] Registering candidate plugin; module='syslogformat', context='parser', name='syslog-parser'
[2018-03-07T16:59:24.741519] Registering candidate plugin; module='csvparser', context='parser', name='csv-parser'
[2018-03-07T16:59:24.741523] Registering candidate plugin; module='afsocket', context='source', name='unix-stream'
[2018-03-07T16:59:24.741527] Registering candidate plugin; module='afsocket', context='destination', name='unix-stream'
[2018-03-07T16:59:24.741531] Registering candidate plugin; module='afsocket', context='source', name='unix-dgram'
[2018-03-07T16:59:24.741545] Registering candidate plugin; module='afsocket', context='destination', name='unix-dgram'
[2018-03-07T16:59:24.741550] Registering candidate plugin; module='afsocket', context='source', name='tcp'
[2018-03-07T16:59:24.741554] Registering candidate plugin; module='afsocket', context='destination', name='tcp'
[2018-03-07T16:59:24.741568] Registering candidate plugin; module='afsocket', context='source', name='tcp6'
[2018-03-07T16:59:24.741581] Registering candidate plugin; module='afsocket', context='destination', name='tcp6'
[2018-03-07T16:59:24.741585] Registering candidate plugin; module='afsocket', context='source', name='udp'
[2018-03-07T16:59:24.741589] Registering candidate plugin; module='afsocket', context='destination', name='udp'
[2018-03-07T16:59:24.741593] Registering candidate plugin; module='afsocket', context='source', name='udp6'
[2018-03-07T16:59:24.741597] Registering candidate plugin; module='afsocket', context='destination', name='udp6'
[2018-03-07T16:59:24.741601] Registering candidate plugin; module='afsocket', context='source', name='syslog'
[2018-03-07T16:59:24.741605] Registering candidate plugin; module='afsocket', context='destination', name='syslog'
[2018-03-07T16:59:24.741613] Registering candidate plugin; module='afsocket', context='source', name='network'
[2018-03-07T16:59:24.741617] Registering candidate plugin; module='afsocket', context='destination', name='network'
[2018-03-07T16:59:24.741621] Registering candidate plugin; module='afsocket', context='source', name='systemd-syslog'
[2018-03-07T16:59:24.741625] Registering candidate plugin; module='cryptofuncs', context='template-func', name='uuid'
[2018-03-07T16:59:24.741629] Registering candidate plugin; module='cryptofuncs', context='template-func', name='hash'
[2018-03-07T16:59:24.741634] Registering candidate plugin; module='cryptofuncs', context='template-func', name='sha1'
[2018-03-07T16:59:24.741638] Registering candidate plugin; module='cryptofuncs', context='template-func', name='sha256'
[2018-03-07T16:59:24.741642] Registering candidate plugin; module='cryptofuncs', context='template-func', name='sha512'
[2018-03-07T16:59:24.741646] Registering candidate plugin; module='cryptofuncs', context='template-func', name='md4'
[2018-03-07T16:59:24.741650] Registering candidate plugin; module='cryptofuncs', context='template-func', name='md5'
[2018-03-07T16:59:24.741654] Registering candidate plugin; module='graphite', context='template-func', name='graphite_output'
[2018-03-07T16:59:24.741659] Registering candidate plugin; module='xml', context='parser', name='xml'
[2018-03-07T16:59:24.741663] Registering candidate plugin; module='afstomp', context='destination', name='stomp'
[2018-03-07T16:59:24.741668] Registering candidate plugin; module='disk-buffer', context='inner-dest', name='disk_buffer'
[2018-03-07T16:59:24.741672] Registering candidate plugin; module='afprog', context='source', name='program'
[2018-03-07T16:59:24.741675] Registering candidate plugin; module='afprog', context='destination', name='program'
[2018-03-07T16:59:24.741680] Registering candidate plugin; module='system-source', context='source', name='system'
[2018-03-07T16:59:24.741684] Registering candidate plugin; module='kvformat', context='parser', name='kv-parser'
[2018-03-07T16:59:24.741688] Registering candidate plugin; module='kvformat', context='parser', name='linux-audit-parser'
[2018-03-07T16:59:24.741692] Registering candidate plugin; module='kvformat', context='template-func', name='format-welf'
[2018-03-07T16:59:24.741695] Registering candidate plugin; module='map-value-pairs', context='parser', name='map_value_pairs'
[2018-03-07T16:59:24.741700] Registering candidate plugin; module='basicfuncs', context='template-func', name='grep'
[2018-03-07T16:59:24.741705] Registering candidate plugin; module='basicfuncs', context='template-func', name='if'
[2018-03-07T16:59:24.741708] Registering candidate plugin; module='basicfuncs', context='template-func', name='or'
[2018-03-07T16:59:24.741711] Registering candidate plugin; module='basicfuncs', context='template-func', name='context-lookup'
[2018-03-07T16:59:24.741716] Registering candidate plugin; module='basicfuncs', context='template-func', name='context-length'
[2018-03-07T16:59:24.741720] Registering candidate plugin; module='basicfuncs', context='template-func', name='context-values'
[2018-03-07T16:59:24.741725] Registering candidate plugin; module='basicfuncs', context='template-func', name='echo'
[2018-03-07T16:59:24.741728] Registering candidate plugin; module='basicfuncs', context='template-func', name='length'
[2018-03-07T16:59:24.741732] Registering candidate plugin; module='basicfuncs', context='template-func', name='substr'
[2018-03-07T16:59:24.741736] Registering candidate plugin; module='basicfuncs', context='template-func', name='strip'
[2018-03-07T16:59:24.741740] Registering candidate plugin; module='basicfuncs', context='template-func', name='sanitize'
[2018-03-07T16:59:24.741744] Registering candidate plugin; module='basicfuncs', context='template-func', name='lowercase'
[2018-03-07T16:59:24.741748] Registering candidate plugin; module='basicfuncs', context='template-func', name='uppercase'
[2018-03-07T16:59:24.741753] Registering candidate plugin; module='basicfuncs', context='template-func', name='replace-delimiter'
[2018-03-07T16:59:24.741757] Registering candidate plugin; module='basicfuncs', context='template-func', name='padding'
[2018-03-07T16:59:24.741761] Registering candidate plugin; module='basicfuncs', context='template-func', name='binary'
[2018-03-07T16:59:24.741765] Registering candidate plugin; module='basicfuncs', context='template-func', name='dirname'
[2018-03-07T16:59:24.741769] Registering candidate plugin; module='basicfuncs', context='template-func', name='basename'
[2018-03-07T16:59:24.741773] Registering candidate plugin; module='basicfuncs', context='template-func', name='list-concat'
[2018-03-07T16:59:24.741778] Registering candidate plugin; module='basicfuncs', context='template-func', name='list-head'
[2018-03-07T16:59:24.741782] Registering candidate plugin; module='basicfuncs', context='template-func', name='list-nth'
[2018-03-07T16:59:24.741786] Registering candidate plugin; module='basicfuncs', context='template-func', name='list-tail'
[2018-03-07T16:59:24.741790] Registering candidate plugin; module='basicfuncs', context='template-func', name='list-slice'
[2018-03-07T16:59:24.741795] Registering candidate plugin; module='basicfuncs', context='template-func', name='list-count'
[2018-03-07T16:59:24.741799] Registering candidate plugin; module='basicfuncs', context='template-func', name='list-append'
[2018-03-07T16:59:24.741803] Registering candidate plugin; module='basicfuncs', context='template-func', name='+'
[2018-03-07T16:59:24.741807] Registering candidate plugin; module='basicfuncs', context='template-func', name='-'
[2018-03-07T16:59:24.741811] Registering candidate plugin; module='basicfuncs', context='template-func', name='*'
[2018-03-07T16:59:24.741817] Registering candidate plugin; module='basicfuncs', context='template-func', name='/'
[2018-03-07T16:59:24.741821] Registering candidate plugin; module='basicfuncs', context='template-func', name='%'
[2018-03-07T16:59:24.741827] Registering candidate plugin; module='basicfuncs', context='template-func', name='sum'
[2018-03-07T16:59:24.741831] Registering candidate plugin; module='basicfuncs', context='template-func', name='min'
[2018-03-07T16:59:24.741835] Registering candidate plugin; module='basicfuncs', context='template-func', name='max'
[2018-03-07T16:59:24.741840] Registering candidate plugin; module='basicfuncs', context='template-func', name='average'
[2018-03-07T16:59:24.741844] Registering candidate plugin; module='basicfuncs', context='template-func', name='ipv4-to-int'
[2018-03-07T16:59:24.741848] Registering candidate plugin; module='basicfuncs', context='template-func', name='indent-multi-line'
[2018-03-07T16:59:24.741852] Registering candidate plugin; module='basicfuncs', context='template-func', name='env'
[2018-03-07T16:59:24.741857] Registering candidate plugin; module='basicfuncs', context='template-func', name='template'
[2018-03-07T16:59:24.741862] Registering candidate plugin; module='dbparser', context='parser', name='db-parser'
[2018-03-07T16:59:24.741883] Registering candidate plugin; module='dbparser', context='parser', name='grouping-by'
[2018-03-07T16:59:24.741888] Registering candidate plugin; module='afuser', context='destination', name='usertty'
[2018-03-07T16:59:24.741892] Registering candidate plugin; module='appmodel', context='root', name='application'
[2018-03-07T16:59:24.741897] Registering candidate plugin; module='appmodel', context='parser', name='app-parser'
[2018-03-07T16:59:24.741901] Registering candidate plugin; module='snmptrapd-parser', context='parser', name='snmptrapd-parser'
[2018-03-07T16:59:24.741905] Registering candidate plugin; module='cef', context='template-func', name='format-cef-extension'
[2018-03-07T16:59:24.741909] Registering candidate plugin; module='date', context='parser', name='date-parser'
[2018-03-07T16:59:24.741913] Registering candidate plugin; module='add-contextual-data', context='parser', name='add_contextual_data'
[2018-03-07T16:59:24.741919] Registering candidate plugin; module='pseudofile', context='destination', name='pseudofile'
[2018-03-07T16:59:24.742516] Finishing include; content='parser generator app-parser', depth='2'
[2018-03-07T16:59:24.742561] Finishing include; content='parser generator app-parser', depth='2'
[2018-03-07T16:59:24.742585] Finishing include; content='source generator system', depth='1'
[2018-03-07T16:59:24.743424] Compiling #unnamed sequence [log] at [/etc/syslog-ng/syslog-ng.conf:4:7]
[2018-03-07T16:59:24.743433]   Compiling src reference [source] at [/etc/syslog-ng/syslog-ng.conf:4:7]
[2018-03-07T16:59:24.743438]     Compiling src sequence [source] at [/etc/syslog-ng/syslog-ng.conf:2:1]
[2018-03-07T16:59:24.743440]       Compiling #unnamed junction [log] at [/etc/syslog-ng/syslog-ng.conf:2:13]
[2018-03-07T16:59:24.743443]         Compiling #unnamed sequence [log] at [source generator system:2:5]
[2018-03-07T16:59:24.743446]           Compiling #unnamed sequence [source] at [source generator system:2:5]
[2018-03-07T16:59:24.743448]             Compiling #unnamed junction [log] at [source generator system:2:13]
[2018-03-07T16:59:24.743451]               Compiling #unnamed sequence [log] at [source generator system:4:5]
[2018-03-07T16:59:24.743454]                 Compiling #unnamed sequence [source] at [source generator system:4:5]
[2018-03-07T16:59:24.743457]                   Compiling #unnamed junction [log] at [source generator system:4:13]
[2018-03-07T16:59:24.743459]                     Compiling #unnamed single [log] at [source generator system:4:14]
[2018-03-07T16:59:24.743464]                 Compiling #unnamed sequence [rewrite] at [source generator system:6:5]
[2018-03-07T16:59:24.743467]                   Compiling #unnamed single [log] at [source generator system:6:15]
[2018-03-07T16:59:24.743469]               Compiling #unnamed single [log] at [source generator system:8:1]
[2018-03-07T16:59:24.743472]           Compiling #unnamed junction [log] at [source generator system:11:1]
[2018-03-07T16:59:24.743474]             Compiling #unnamed sequence [log] at [source generator system:11:10]
[2018-03-07T16:59:24.743476]               Compiling #unnamed junction [log] at [source generator system:12:3]
[2018-03-07T16:59:24.743478]                 Compiling #unnamed sequence [log] at [source generator system:13:5]
[2018-03-07T16:59:24.743480]                   Compiling #unnamed sequence [parser] at [source generator system:13:5]
[2018-03-07T16:59:24.743483]                     Compiling #unnamed sequence [log] at [parser generator app-parser:3:5]
[2018-03-07T16:59:24.743485]                       Compiling #unnamed junction [log] at [parser generator app-parser:3:5]
[2018-03-07T16:59:24.743487]                     Compiling #unnamed sequence [log] at [parser generator app-parser:3:5]
[2018-03-07T16:59:24.743489]                       Compiling #unnamed junction [log] at [parser generator app-parser:3:5]
[2018-03-07T16:59:24.743493]                 Compiling #unnamed sequence [log] at [source generator system:19:12]
[2018-03-07T16:59:24.743498]         Compiling #unnamed single [log] at [/etc/syslog-ng/syslog-ng.conf:2:24]
[2018-03-07T16:59:24.743502]   Compiling syslog_sygic reference [destination] at [/etc/syslog-ng/syslog-ng.conf:4:20]
[2018-03-07T16:59:24.743504]     Compiling syslog_sygic sequence [destination] at [/etc/syslog-ng/syslog-ng.conf:3:1]
[2018-03-07T16:59:24.743508]       Compiling #unnamed junction [log] at [/etc/syslog-ng/syslog-ng.conf:3:27]
[2018-03-07T16:59:24.743511]         Compiling #unnamed single [log] at [/etc/syslog-ng/syslog-ng.conf:3:28]
[2018-03-07T16:59:24.743928] Error setting up TLS session context; tls_error='(null):(null):(null)'
[2018-03-07T16:59:24.743955] Error initializing message pipeline; plugin name='network', location='/etc/syslog-ng/syslog-ng.conf:3:28'
````

**Comment by MrAnno** [comment]
@hydrapolic Thanks again.

The problem here is that you forgot to specify your TLS key and cert file (`key-file()` and `cert-file()`).
As an alternative, you can use a PKCS12 file as well (`pkcs12-file()`).

It might be possible that syslog-ng could be started with this configuration in previous releases, but the destination didn't funtioned properly in such cases.

The error message `tls_error='(null):(null):(null)` is pretty useless , I'm going to open a pull request with error logging improvements.

**Comment by hydrapolic** [comment]
Hello @MrAnno,
this configuration sets up a client that connects to a TLS enabled server with a regular certificate (not self-signed). Syslog-ng should not use any client certificate, so we only verify the server, not the client.

When I checked the network, the messages were encrypted and fiddling the trusted-keys rendered the logging to stop. So from that I assumed it works properly.

It's also mentioned here:
https://syslog-ng.com/documents/html/syslog-ng-ose-latest-guides/en/syslog-ng-ose-guide-admin/html/procedure-configuring-tls-client.html

**Comment by MrAnno** [comment]
@hydrapolic 
Sorry, you are absolutely right.

It's a bug introduced by 33a61f9237ade38eb19f80dfd597c3bef330ee96. key-file and cert-file are optional in case of a destination.

TODO:
- improve error logging in TLSContext
- fix `tls_context_load_key_and_cert` to make key and cert files optional again

**Comment by MrAnno** [comment]
This issue will be fixed in #1917.

@hydrapolic Thank you for letting us know about this :+1:

**Comment by hydrapolic** [comment]
Thank you for a such a quick response, tested locally, it works fine :)
